src/lib/state/DY.Lib.State.Database.fst

```diff
@@ -0,0 +1,413 @@
+  ;
+  db_global_pred_update: tr1:trace -> tr2:trace -> prin:principal -> sess_id:state_id -> db1:list row_t -> db2:list row_t -> Lemma
+    (requires db_global_pred tr1 prin sess_id db1 /\
+      tr1 <$ tr2 /\
+      list_growsP (row_update_pred tr2 prin) db1 db2
+    )
+    (ensures db_global_pred tr2 prin sess_id db2)
+  ;
+  // No knowable lemma needed for the overall database, because it holds by construction.
+}
+
+/// TODO give actual instances
+let ps_state_id : parser_serializer bytes state_id = admit()
+let ps_string : parser_serializer bytes string = admit()
+
+instance parseable_serializeable_bytes_db_row (row_t:Type0) {|db_t:db_types row_t|} : parseable_serializeable bytes row_t = mk_parseable_serializeable (db_t.ps_row_t)
+
+instance local_state_db_entry (row_t:Type0) {|db_t:db_types row_t|}: local_state row_t = {
+  tag = db_t.row_tag;
+  format = (parseable_serializeable_bytes_db_row row_t);
+}
+
+/// TODO: Could also make an indirection record type to the state ID, making it easier to add more metadata
+[@@ with_bytes bytes]
+noeq type db = {
+  [@@@ with_parser #bytes (ps_list (ps_state_id))]
+  rows: list (state_id)
+}
+
+%splice [ps_db] (gen_parser (`db))
+%splice [ps_db_is_well_formed] (gen_is_well_formed_lemma (`db))
+
+instance parseable_serializeable_bytes_db : parseable_serializeable bytes db = mk_parseable_serializeable (ps_db)
+
+instance local_state_db (row_t:Type0) {|db_t:db_types row_t|}: local_state db = {
+  tag = db_t.db_tag;
+  format = (parseable_serializeable_bytes_db);
+}
+
+val db_row_session_invariant:
+  {|crypto_invariants|} ->
+  #row_t:Type0 -> {|db_types row_t|} ->
+  db_pred:db_predicate row_t ->
+  local_state_predicate row_t
+let db_row_session_invariant #cinvs #row_t #db_t db_pred = {
+  pred = (fun tr prin state_id row -> db_pred.row_pred tr prin row);
+  pred_later = (fun tr1 tr2 prin state_id row -> db_pred.row_pred_later tr1 tr2 prin row);
+  pred_knowable = (fun tr prin state_id row ->
+    db_pred.row_pred_knowable tr prin row;
+    assert(is_well_formed _ (is_knowable_by (principal_label prin) tr) row);
+    assert(principal_label prin `can_flow tr` principal_state_label prin state_id);
+//    assert(is_well_formed _ (is_knowable_by (principal_state_label prin state_id) tr) row);
+    ()
+  );
+}
+
+
+val is_pointer_to_row :
+  #row_t:Type0 -> {|db_types row_t|} ->
+  tr:trace ->
+  prin:principal -> row_state_id:state_id ->
+  prop
+let is_pointer_to_row #row_t #db_t tr prin row_state_id =
+  let (row_opt, _) = get_state #row_t prin row_state_id tr in
+  Some? row_opt
+
+val get_row_by_pointer :
+  #row_t:Type0 -> {|db_types row_t|} ->
+  tr:trace ->
+  prin:principal -> row_state_id:state_id{is_pointer_to_row #row_t tr prin row_state_id} ->
+  row_t
+let get_row_by_pointer #row_t #db_t tr prin row_state_id =
+  let (row_opt, _) = get_state #row_t prin row_state_id tr in
+  Some?.v row_opt
+
+#push-options "--fuel 1 --ifuel 1"
+val list_coerceP :
```

Comment:
This feels like a much more generally useful function that I'm surprised isn't part of the core library in some form. Maybe there's a simpler way to accomplish the same thing?

Reply:
The function `FStar.List.Tot.list_ref` might be doing roughly the same thing

Reply:
Indeed, with just an extra call to `for_allP_eq`, it accomplishes the same thing, and gives nicer postconditions.